presto-raptorx/src/main/java/com/facebook/presto/raptorx/metadata/MasterWriterDao.java

```diff
@@ -0,0 +1,323 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.raptorx.metadata;
+
+import org.jdbi.v3.sqlobject.customizer.Bind;
+import org.jdbi.v3.sqlobject.statement.SqlQuery;
+import org.jdbi.v3.sqlobject.statement.SqlUpdate;
+import org.jdbi.v3.sqlobject.statement.UseRowMapper;
+
+import java.util.List;
+
+public interface MasterWriterDao
+{
+    // transaction
+
+    @SqlQuery("SELECT commit_id FROM current_commit FOR UPDATE")
+    Long getLockedCurrentCommitId();
+
+    @SqlUpdate("INSERT INTO active_commit (singleton, commit_id, start_time, rolling_back)\n" +
+            "VALUES ('X', :commitId, :startTime, FALSE)")
+    void insertActiveCommit(
```

Comment:
This is another high level design question.  The active_commit seems to serve the same purpose as the traditional UNDO log. The recover logic is based on it. So may I ask why not go for a traditional UNDO log, on local SSD log file.

Is the reason that mysql table has replicas, so higher reliability?  I guess if our mysql doesn't have a synchronous replica, UNDO log is pretty much the same, right? Logic won't be much different.

Reply:
Yes, this is the undo / rollback log. We use MySQL for everything because implementing a reliable, durable database, even for a single machine, is a huge undertaking. Doing it while accounting for machine failure in a "cloud" type environment is even harder (and gets into all sorts of deployment specific concerns such as persistent disks). We use MySQL for everything because it handles these things for us.